### PR TITLE
Cardinality many from array

### PR DESCRIPTION
**Pull request by mhoeger**
Related to: https://github.com/Azure/azure-functions-dotnet-worker/pull/205 , which implements cardinality many for anything that is an IEnumerable. This is scoped down to array types, which is a more simple rule to follow on when data type is batched vs. isn't.

Also appears consistent with the current WebJobs contract for "cardinality": "many"
https://github.com/Azure/azure-functions-host/blob/v3.0.15417/src/WebJobs.Script/Binding/GeneralScriptBindingProvider.cs#L72-L77

Resolves https://github.com/Azure/azure-functions-dotnet-worker/issues/127
Resolves https://github.com/Azure/azure-functions-dotnet-worker/issues/115

### SDK Changes
- For inputs, try to set "dataType" if we know the requested payload is `string` or `byte[]`
- Try set "cardinality": "many" if we know that the requested payload implements IEnumerable or IEnumerable Generic
  - Set "dataType" if we know that the requested payload inside of the collection is `string` or `byte[]`

### .NET Worker Changes
- Enable capability to receive TypedData of:
  - collection bytes
  - collection string
  - collection double
  - collection long
- New converter which converts `IEnumerable<string>`, `IEnumerable<byte[]>`, `IEnumerable<double>`, and `IEnumerable<long>` (from above capability) to Array type
- _Observation: not all types the user requests (ex: "Queue" which implements IEnumerable) will be properly converted, although "cardinality": "many" will be set. Potentially worth surfacing some sort of warning or error if we are unable to convert. From user perspective seems difficult to debug._

### CODE REVIEW

File: sdk/Sdk/FunctionMetadataGenerator.cs [hunk 1]
```diff
@@ -462,5 +487,55 @@ private static bool TryGetBaseAttributeType(CustomAttribute attribute, string ba
...
+            {
+                return DataType.Binary;
+            }
+
+            return DataType.Undefined;
+        }
+
+        private static bool IsStringType(string fullName)
+        {
+            return string.Equals(fullName, typeof(string).FullName, StringComparison.Ordinal);
```
Comment (fabiocav) [inline sdk/Sdk/FunctionMetadataGenerator.cs:525]: nit: but as an optimization, the typeof(<type>) name can be resolved and reused.

File: src/DotNetWorker/Converters/EnumerableConverter.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            if (context.Parameter.Type.IsArray)
+            {
+                targetType = EnumerableTargetType.Array;
+            }
+
+            // Only apply if user is requesting an array, list, or hashset
+            if (targetType is not null)
+            {
+                // Valid options from FunctionRpc.proto are string, byte, double and long collection
+                if (context.Source is IEnumerable<string> enumerableString)
```
Comment (fabiocav) [inline src/DotNetWorker/Converters/EnumerableConverter.cs:26]: Would this be a little simpler using pattern matching on the source?

File: src/DotNetWorker/Converters/EnumerableConverter.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                    return true;
+                }
+            }
+
+            target = default;
+            return false;
+        }
+
+        // Dictionary and Lookup not handled because we don't know 
+        // what they keySelector and elementSelector should be.
```
Comment (fabiocav) [inline src/DotNetWorker/Converters/EnumerableConverter.cs:53]: Can you clarify this?

File: src/DotNetWorker/Converters/EnumerableConverter.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            switch (targetType)
+            {
+                case EnumerableTargetType.Array:
+                    return source.ToArray();
+                default:
+                    return null;
+            }
+        }
+
+        private enum EnumerableTargetType
```
Comment (ankitkumarr) [inline src/DotNetWorker/Converters/EnumerableConverter.cs:65]: Wondering if this is needed, as I see after the recent change to move to only Array type, this file only ever uses the `Array` enum value.

**Review by ankitkumarr — APPROVED** [review]
Looks great to me! Just a couple of optional nits.

**Comment by mhoeger** [comment]
Closing in favor of - https://github.com/Azure/azure-functions-dotnet-worker/pull/205
We won't be doing any magic sniffing because array type != batched events

**Pull request closed without merging**